src/coreclr/vm/eventing/eventpipe/ds-rt-coreclr.h

```diff
@@ -329,6 +330,45 @@ ds_rt_disable_perfmap (void)
 #endif // FEATURE_PERFMAP
 }
 
+static
+uint32_t
+ds_rt_appcontext_properties_get (dn_vector_ptr_t *props_array)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (props_array != NULL);
+
+	return Configuration::EnumerateKnobs([](const LPCWSTR& name, const LPCWSTR& value, void* context) {
```

Comment:
Are there any properties that we might want to omit from here?

Reply:
Here's the list for a very basic .NET 8 console app:
- FX_DEPS_FILE
- TRUSTED_PLATFORM_ASSEMBLIES
- NATIVE_DLL_SEARCH_DIRECTORIES
- PLATFORM_RESOURCE_ROOTS
- APP_CONTEXT_BASE_DIRECTORY
- APP_CONTEXT_DEPS_FILES
- PROBING_DIRECTORIES
- RUNTIME_IDENTIFIER
- System.Reflection.Metadata.MetadataUpdater.IsSupported
- HOST_RUNTIME_CONTRACT
- STARTUP_HOOKS

Other identifiers are added to this list when the app contains them in the `runtimeOptions:configProperties` section in the *.runtimeconfig.json file.

Probably HOST_RUNTIME_CONTRACT can be skipped because its value is a in-memory address to the host runtime contract. Everything else seems to be directory and file paths of various types.

My default position would be to only filter out HOST_RUNTIME_CONTRACT and leave everything else in in the list for reading values. In a future change that may allow modification of values, I can see some more of these being effectively made read-only, but I would have to defer to others on what's desirable to allowed to be read.

@vitek-karas and @elinor-fung, do you have suggestions on what to filter (if any and where to do it) for this initial PR that enables reading the app config values?

Reply:
The total size of all these switches can be in 10's kB. I assume that this event is going to be typically enabled by default. Do you really want to send 10's kB over the diagnostic channel during startup by default?

I would ask a different question. What do you expect that this data is going to be used for?

Reply:
> The total size of all these switches can be in 10's kB. I assume that this event is going to be typically enabled by default. Do you really want to send 10's kB over the diagnostic channel during startup by default?

There's already an existing diagnostic command for collecting the entire environment block of the target process. I would hazard a guess that it's has similar size characteristics. We haven't seen any significant slowdown during startup nor have we received any customer feedback about it.

The only one that I see right now that is significantly large is `TRUSTED_PLATFORM_ASSEMBLIES`, which we could probably do without. Maybe we can categorize the data and have those categories as part of the request payload on the diagnostic command in order to limit the response to the interested data.

> I would ask a different question. What do you expect that this data is going to be used for?

Originally, we only wanted to get startup hook information, and append our own startup hook to the list which the process is held up during the diagnostics suspension (which occurs just before managed code execution is started); this request is #83756. It was suggested that this be generalized to get / set app context properties. Expanding it to this general concept also allows us to:
- read `RUNTIME_IDENTIFIER` to determine what type of `ICorProfilerCallback` implementation we should load in the target process, especially differentiating on glibc vs musl-libc implementations.

The following are hypothetical but likely candidates for use:
- read `System.StartupHookProvider.IsSupported` and determine if we should bother with the startup hook at all. This would also allow us to automatically tell users if features that depend on the startup hook will work. Other app context switches may be useful; not sure how to determine the full set of "in box" switches though.
- read `BUNDLE_PROBE` \ `HOSTPOLICY_EMBEDDED` \ `PINVOKE_OVERRIDE` to check if the process is single file; some diagnostic scenarios do not work in single file applications. It would be good to know whether we bother trying to set up for those scenarios or be able to give the user a better indication why the diagnostic scenario didn't work (e.g. "sorry, this feature is not compatible with single file applications").
- read `APP_CONTEXT_BASE_DIRECTORY` and tell our profiler what that is; this may help in discovering application symbol information that was published with the application so that we can report that back to the .NET Monitor process in order to get SourceLink information. We can use that information in combination with diagnostic artifacts that are collected to help guide users or automated tooling back to real source code.
- (this is a stretch) read / set `APP_PATHS` / `PROBING_DIRECTORIES` to get our managed assemblies loadable without having to rely on managed assembly resolving and LoadFrom context.

Those are the obvious ones that I can think of right now; I can probably dig out more scenarios where some of this data may be useful.

Reply:
For my own reference, some (all?) feature switches seem to be documented [here](https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md) and surfaced in the SDK [here](https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L446).

Reply:
> We haven't seen any significant slowdown during startup nor have we received any customer feedback about it.

In last year, I have experienced myself visible slowdowns from excesive default diagnostic tracing and I have seens others run into performance impact of excessive diagnostic tracing as well. The cost of the default diagnostic tracing is something that I have started to keep an eye on.

> read RUNTIME_IDENTIFIER to determine what type of ICorProfilerCallback implementation we should load in the target process, especially differentiating on glibc vs musl-libc implementations.

Note that `RUNTIME_IDENTIFIER` is not 100% reliable indicator of the glibc library type. It allows you to guess and the guess maybe correct 99%+ of the time, but it is not 100% reliable. If you want to know libc library name, you should let the runtime log it so you do not need to guess.

> The following are hypothetical but likely candidates for use:

Would it be better to do most of this or even all this in your profiler code or startup hook in-proc? It feels like this tracing is just a workaround for the actual problem.

> For my own reference, some (all?) feature switches seem to be documented 

These are .NET runtime switches. NuGet packages can define their own.

Reply:
> The total size of all these switches can be in 10's kB. I assume that this event is going to be typically enabled by default. Do you really want to send 10's kB over the diagnostic channel during startup by default?
> 
> I would ask a different question. What do you expect that this data is going to be used for?

This is an IPC command and not an event, it is returned over the IPC channel when an IPC client requests it so it shouldn't affect startup unless the implementation of the command requires additional runtime bookkeeping or other support service to be enabled (I don't see anything like that in this PR). IPC commands are triggered on demand by a IPC channel client that needs the information. It is not part of the regular EventPipe event stream since its not an event.

Reply:
> In last year, I have experienced myself visible slowdowns from excesive default diagnostic tracing and I have seens others run into performance impact of excessive diagnostic tracing as well. The cost of the default diagnostic tracing is something that I have started to keep an eye on.

That feels like a generalization about diagnostic tooling. I'm not certain how invoking this command in the PR once automatically and maybe successively on demand has anything to do with tracing. If you have a specific scenario or observance of the .NET Monitor tool causing unacceptable performance degradation, please log an issue in https://github.com/dotnet/dotnet-monitor and we will investigate it.

> Note that `RUNTIME_IDENTIFIER` is not 100% reliable indicator of the libc library type. It allows you to guess and the guess maybe correct 99%+ of the time, but it is not 100% reliable. If you want to know libc library name, you should let the runtime log it so you do not need to guess.

If we can't get #74476 for .NET 8, I would gladly take "correct 99%+ of the time". I think this falls under "not letting perfect be and enemy of the good."

> Would it be better to do most of this or even all this in your profiler code or startup hook in-proc? It feels like this tracing is just a workaround for the actual problem.

I would rather .NET Monitor not involve the profiler implementation unless the user opts into a feature that requires it, and not all in-process .NET Monitor features require the profiler. Additionally, profilers can be heavy-handed, especially in production systems. As for each idea that I presented:
- read `System.StartupHookProvider.IsSupported`: to my knowledge, that's not doable from a profiler EXCEPT if it were to ReJIT the `StartupHookProvider.get_IsSupported` method and then call back into some custom managed implementation somewhere (either emitted by the profiler or reflection invoked to assembly load some helper assembly). All of that is very expensive to do at runtime, is very expensive from engineering to get right at compile time, and can cause invalid programs if not authored correctly. For all we know, that gets trimmed out for trimmed applications, so our ReJIT'd methods would not get invoked. Finally, there's no feedback that we can provide to the user if we are only able to indeterminately calculate this from within the process. If there's any other way to look up that information, I would prefer it do it that way and leave this as absolute last resort.
- read `BUNDLE_PROBE` \ `HOSTPOLICY_EMBEDDED` \ `PINVOKE_OVERRIDE`: to my knowledge, that's not doable from a profiler either. There is no indication that an application is deployed as a single file application.
- read `APP_CONTEXT_BASE_DIRECTORY`: this one is doable with a profiler by enumerating modules and listening for loads. However, if we do not have any other features that require listening for module loads, I would not want to opt into this behavior. Additionally, scanning for symbols on-demand or in background is probably preferable to performing that on each module load.
- read / set `APP_PATHS` / `PROBING_DIRECTORIES`: this too is not doable with a profiler without going through ReJIT of specific methods, with callbacks, emitting assemblies and methods, etc...

Today, the only thing the .NET Monitor profiler is capable of doing is enumerating threads and do stack walks on an on-demand basis in order to capture call stacks in leu of capturing a full dump of the process; other features may come online in the future, but presently, the current profiler implementation is fairly non-invasive.

The .NET Monitor tool is designed to be a production agent for capturing diagnostic artifacts and information from explicit requests via its HTTP API or from automated collection rules and triggers; we will try to do anything we can to automatically configure diagnostic capabilities (such as setting the "arbitrary" startup hook path or determining the platform/libc/architecture of the target process) so that the user doesn't have to explicitly configure these. There is very little room for user interactivity to fix misconfiguration and, in some cases (such as setting startup hooks to a path that doesn't exist), misconfiguration is catastrophic for the target application; minimizing the need for the user to understand technical aspects or the implementation details of how paths to assemblies are constructed greatly increases the probability of success in setting up diagnostics.

Reply:
Re list of properties passed from the host:
* There's no master list - there's what the host populates by default, but it's considered an implementation detail between the host and the runtime.
* There are already some properties which are "on demand" and won't be in the list. For example we've moved the bundle probe from a top-level property to a field in the host contract.
* The plan is to migrate as many as possible to become "on demand" for performance reasons.

With this PR you're making the Configuration class the defacto "source of truth" for all of these properties. Maybe that's OK, but it should be very intentional. For example we typically treat `AppContext` as the source of truth for anything in the managed world.

Based on some of the comments above, it seems you're trying to use some of the properties to determine certain characteristics about the app in question. I think it would be good to have a discussion on what are the characteristics you're looking for and how to expose that information from the runtime.
As an example: why do you need to know if the app is single-file? Single-file is a very fuzzy thing - there's the CoreCLR single-file, but then there's NativeAOT which is also effectively single-file, and for example Android apps when packaged are also effectively single-file. So looking at the bundle-probe property will be very problematic, as not all of these will have that property.

Reply:
> With this PR you're making the Configuration class the defacto "source of truth" for all of these properties. Maybe that's OK, but it should be very intentional. For example we typically treat `AppContext` as the source of truth for anything in the managed world.

In my brief exploration of this code, it seems that the Configuration class is more than just the source of truth for the managed world; other parts of the native vm use it ([here](https://github.com/search?q=repo%3Adotnet%2Fruntime+Configuration%3A%3AGetKnobDWORDValue&type=code), [here](https://github.com/search?q=repo%3Adotnet%2Fruntime+Configuration%3A%3AGetKnobULONGLONGValue&type=code), [here](https://github.com/search?q=repo%3Adotnet%2Fruntime+Configuration%3A%3AGetKnobStringValue&type=code), [here](https://github.com/search?q=repo%3Adotnet%2Fruntime%20Configuration%3A%3AGetKnobBooleanValue&type=code)), unless you are classifying the vm itself as part of "the managed world".
 
> Based on some of the comments above, it seems you're trying to use some of the properties to determine certain characteristics about the app in question. I think it would be good to have a discussion on what are the characteristics you're looking for and how to expose that information from the runtime.

The original motivation and ask in #83756 is to get and set startup hook paths. Everything else is bonus so long as it makes sense and is appropriate to use that information. 

> As an example: why do you need to know if the app is single-file? Single-file is a very fuzzy thing - there's the CoreCLR single-file, but then there's NativeAOT which is also effectively single-file, and for example Android apps when packaged are also effectively single-file. So looking at the bundle-probe property will be very problematic, as not all of these will have that property.

For this particular lookup (is it single-file or not), one of the diagnostic problems with single-file applications is that you cannot capture a dump of it using the dump command on the diagnostic IPC channel; this command requires the createdump utility but that is not bundled with single-file apps nor is linked into the single-file host. This bit us (https://github.com/dotnet/dotnet-monitor/issues/2932) when the sample Docker images were updated to publish as single-file (see https://github.com/dotnet/dotnet-docker/issues/4271) and customers started reporting that .NET Monitor couldn't capture dumps anymore from the image. We ended up undoing the single-file publishing in those images to support this very common diagnostic capability. Knowing that the target application is single-file would have made it easy for .NET Monitor to report that the dump capture possibly failed due to single-file target and is not supported at that time.

This is supposedly partially fixed as of last week (see https://github.com/dotnet/runtime/issues/68676) but may not be complete yet (says that the static library was created but not linked in yet).

Anyway, this was an example of what the data _could_ be used for, not a hard dependency on which the successful execution of a capability is dependent. The only ones that we currently need are to be able to add our startup hook (#83756, which requires the ability to read, append, and then set the value in the list that would be eventually passed to the AppContext) and to be able to get the runtime identifier (#74476).

If scaling this back to the original ask would fit for .NET 8, I would gladly do that. Nothing says the more expanded capability could be added in the future again.

Reply:
Re single-file crash dump: Wouldn't it be cleaner if the runtime responded via the IPC channel if the dump could be collected or not? After all, the runtime knows the answer... using hints like single-file is just guessing. I understand that guessing is all we have right now, but if we're introducing new functionality why not design it for the original purpose?

Reply:
> Yes, this is a general concern. We have a combination of two trends: always on diagnostics (this is difference from the past when the diagnostics was typically opt-in), adding diagnostic without thinking much about performance overhead and reliablity (it was ok to skim over this in the past when the diagnostics was opt-in).

.NET Monitor only enables collection of event counters from a small set of providers by default over the event pipe (e.g. tracing) in order to report metrics; this can also be opt-out fairly easily. All other features are opt-in by the user. We do perf investigation for these features and provide guidance for how to use them appropriately and what the anticipated result is if they go overboard on enabling too much. While your concern is warranted as a general principle, in this case, I don't think it is applicable unless you have specific data that shows .NET Monitor causing overhead and reliability issues.

For this PR, this is adding a diagnostic IPC command and has no bearing on tracing. It is for on-demand instantaneous inspection and is not something that continuously causes any cpu or memory utilization.

> Why is the back-and-forth handshake needed? Can we have a diagnostic command to inject a startup hook instead

To check if the startup hook was already applied. We only want to add it if it doesn't already exist. Now, if we want to push that logic into the runtime itself, I guess that _could_ work. However, it would probably require a lot more locking/interlocking/comparisons just in case two clients invoke the "add startup hook command" at the same time. There's also no way for a client to verify that the startup hook was in fact added via this mechanism unless we can guarantee that invoking the command will in fact add it despite any race conditions involving other clients trying to add their startup hooks.

> (that can be even issued in attach scenario - it does not need to be limited to startup only)?

Startup hooks only execute at startup... they are never evaluated again after the first processing of the startup hooks at the very beginning of managed execution. Attach would have to use some other mechanism (e.g. profiler attach + hosting apis to load managed assembly). Unless you are proposing that the invocation of this diagnostic IPC command after startup has executed would do that work itself (e.g. load the managed assembly, invoke the StartupHook.Initialize method).

These last two questions feel like they should be discussed on #83756 rather than in this PR.

Reply:
> For example, you do not have to depend on guessing to figure out what kind of libc is loaded based on the RID.

Determining the libc type during diagnostic suspension is not for startup hook usage; it's for determining which variant of our profiler to use.

For loading the profiler, either AddStartupProfiler and AttachProfiler diagnostic commands are invoked, both of which take a full file path to the profiler implementation. The former command is only usable during the suspension and does not evaluate whether the library is loadable or not because the actual load comes later after the runtime has been resumed. There is no feedback on whether the library is loaded or not, so we can't even just try one and fallback to the other; we need to know the RID before the runtime is resumed. For the latter, we could just try one and then the other if the first fails, but that feels like a waste if we already know what the RID is due to needing it for AddStartupProfiler.

> You do not need to depend on runtime adding built-in tracing for every piece of information that you need.

I have never asked for that. There is no tracing involved.

To make decisions on what needs to be loaded or enable at diagnostic suspension, we need diagnostic commands to get specific information (such as RID) because some features (such as profilers) require native-specific information and managed code is not executing at that time thus cannot help in a platform agnostic way.

Reply:
> we need diagnostic commands to get specific information (such as RID) because some features (such as profilers)

RID is not appropriate for this need. For example, what you are going to do when you see RID `void-x64`?

If you need libc type to load the right profiler flavor, we should introduce diagnostic command that returns the libc type.

Alternatively, you can find out the libc that the process has loaded using `/proc` file system. Is that an option?

Reply:
> If you need libc type to load the right profiler flavor, we should introduce diagnostic command that returns the libc type.

I logged #74476 specifically for this a 9 months ago. I'll try to get to this if the diagnostics team cannot.

> Alternatively, you can find out the libc that the process has loaded using `/proc` file system. Is that an option?

No, .NET Monitor is typically not running in the same process namespace in its primary scenario (multi-container environments such as Kubernetes, where .NET Monitor is typically running as a sidecar container and maybe be running as a different user). We can however use that as a fallback mechanism of almost last resort if it happens to be running in the same namespace, which we may do for runtime versions lower than .NET 8.